src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+    private HTTPRequest readAllLinesAndReturnHttpRequest(ClientReadable bufferedReader)  {
+
+        String character;
+        RequestBuilder requestBuilder = new RequestBuilder();
+        StringBuilder headers = new StringBuilder();
+        StringBuilder body = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
+                        break;
+                    }
+                    body.append(character);
+                    if (body.length() == bodyLength) {
+                        break;
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+                return requestBuilder.build();
+            }
+        }
+        return turnStringBuilderIntoHttpRequest(headers, body, requestBuilder);
+    }
+
+    private HTTPRequest turnStringBuilderIntoHttpRequest(StringBuilder headers, StringBuilder body, RequestBuilder requestBuilder) {
+        String[] headersArray = headers.toString().split(crlf);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
+        }
+        if (body.length() != 0) {
+            requestBuilder.setBody(body.toString());
         }
+        return requestBuilder.build();
     }
+
+    private RequestBuilder discernHeaderTypeAndAddToRequestBuilder(String line, RequestBuilder requestBuilder) {
+        if (!line.contains(": ")) {
+            requestBuilder.buildRequestLine(line);
+        } else {
+            requestBuilder.buildHeaderLine(line);
+
+        }
+        return requestBuilder;
+    }
+
+    private boolean checkIfReachedEndOfHeaders(StringBuilder requestScannedInSoFar) {
+        return requestScannedInSoFar.toString().contains(crlf + crlf);
+    }
+
+    private boolean checkIfHeadersIndicateARequestBody(StringBuilder headers) {
+        String headersAsString = headers.toString();
+        if (headersAsString.contains("Content-Length:")) {
+            if (getBodyLength(headers) == 0) {
```

Comment:
I think you can combine both conditions with `&&` in a single if clause rather than nesting them.

```java
if (headersAsString.contains("Content-Length:") && getBodyLength(headers) == 0) {
    return false
}
return true
```